File: 其他模组/2.9.5/scripts/block/炮塔/暮光.js
const my = require("base/物品");
const lib = require("base/coflib");
const status = require("base/status");
let C8CA9E8 = lib.Color("8CA9E8"), D1EFFF = lib.Color("D1EFFF");
let b = Object.assign(new LaserBulletType(3565), {
    length: 8000,
    width: 40,
    shootEffect: Object.assign(new ParticleEffect(), {
        line: true,
        particles: 12,
        lifetime: 20,
        length: 45,
        cone: 30,
        lenFrom: 6,
        lenTo: 6,
        strokeFrom: 3,
        strokeTo: 0,
        lightColor: C8CA9E8,
        colorFrom: D1EFFF,
        colorTo: C8CA9E8
    }),
    colors: [lib.Color("6569C9"), C8CA9E8, D1EFFF],
    ammoMultiplier: 1,
    status: status.湍能,
    statusDuration: 120,
    sideAngle: 22.5,
    buildingDamageMultiplier: 0.05,
    hitEffect: Object.assign(new ParticleEffect(), {
        line: true,
        particles: 12,
        lifetime: 20,
        length: 75,
        cone: 360,
        lenFrom: 6,
        lenTo: 6,
        strokeFrom: 3,
        strokeTo: 0,
        lightColor: C8CA9E8,
        colorFrom: D1EFFF,
        colorTo: C8CA9E8
    })
})

let mg = lib.SpeedUpTurret(PowerTurret, "暮光", {
    size: 6,
    range: 2160,
    reload: 1200,
    shootType: b,
    shake: 8,
    recoil: 5,
    shootY: 0,
    shootCone: 45,
    rotateSpeed: 2,
    recoilTime: 180,
    cooldownTime: 180,
    canOverdrive: false,
    minWarmup: 0.98,
    shootWarmupSpeed: 0.08,
    warmupMaintainTime: 120,
    fogRadiusMultiuplier: 0.2,
    unitSort: UnitSorts.strongest,
    shootSound: Sounds.malignShoot
}, 120, 3600, 8);
mg.consumePower(165);
mg.consumeLiquid(Liquids.cryofluid, 1.5);
mg.setupRequirements(
    Category.turret,
    BuildVisibility.shown,
    ItemStack.with(
        Items.copper, 3300,
        Items.lead, 2800,
        Items.silicon, 2350,
        my.铱板, 1500,
        my.导能回路, 1900,
        my.陶钢, 900,
        Items.phaseFabric, 585,
        my.肃正协议, 1
    )
);

let pro = DrawPart.PartProgress;
mg.drawer = (() => {
    const d = new DrawTurret();
    for (let i = 0; i < 5; i++) {
        d.parts.add(
            Object.assign(new RegionPart("-aim"), {
                heatProgress: pro.warmup.sin((4 - i) * 10, 20, 1),
                heatLight: true,
                drawRegion: false,
                y: (i + 1) * 24 / (1 - i / 24),
                xScl: 1 - i * 0.1,
                yScl: 1 - i * 0.1,
                heatColor: lib.Color("D1EFFF")
            })
        )
    }
    d.parts.add(
        Object.assign(new RegionPart("-glow"), {
            heatProgress: pro.warmup,
            heatColor: lib.Color("F03B0E"),
            drawRegion: false
        })
    )
    return d;
})();

File: src/ice/content/block/turret/暮光.kt
package ice.content.block.turret


import arc.graphics.Color
import arc.math.Mathf
import arc.util.Time
import ice.content.IItems
import ice.content.ILiquids
import ice.content.IStatus
import ice.entities.bullet.LaserBulletType
import mindustry.entities.UnitSorts
import mindustry.entities.effect.ParticleEffect
import mindustry.entities.part.DrawPart
import mindustry.entities.part.RegionPart
import mindustry.gen.Sounds
import mindustry.type.Category
import mindustry.type.Liquid
import singularity.world.blocks.turrets.SglTurret
import singularity.world.draw.DrawSglTurret

class 暮光 :SglTurret("turret_twilight") {
  init {
    localization {
      zh_CN {
        localizedName = "暮光"
        description = "重型激光炮塔,向指定方向发射一道超视距湍能激光束\n其聚能速度会随聚焦晶体的预热而逐渐提升"
      }
    }
    warmupSpeed=0.02f
    size = 6
    health = 12000
    range = 800f
    shake = 8f
    recoil = 5f
    shootY = 0f
    shootCone = 45f
    rotateSpeed = 2f
    recoilTime = 180f
    cooldownTime = 180f
    canOverdrive = false
    unitSort = UnitSorts.strongest
    shootSound = Sounds.shootMalign
    drawers = DrawSglTurret().apply {
      for(i in 0..4) {
        parts.add(RegionPart("-aim").apply {
          heatProgress = DrawPart.PartProgress{ partParams ->
            partParams.warmup * Mathf.sin(Time.time + (4 - i) * 10f, 20f, 1f)
          }
          heatLight = true
          drawRegion = false
          y = (i + 2) * 24f / (1f - i / 24f)
          xScl = 1f - i * 0.1f
          yScl = 1f - i * 0.1f
          heatColor = Color.valueOf("D1EFFF")
        })
      }

      parts.add(RegionPart("-glow").apply {
        heatProgress = DrawPart.PartProgress.warmup
        heatColor = Color.valueOf("F03B0E")
        drawRegion = false
      })
    }
    newCoolant(1f, 0.4f, { l: Liquid? -> l!!.heatCapacity >= 0.4f && l.temperature <= 0.5f }, 0.25f, 20f)

    requirements(
      Category.turret,
      IItems.铜锭, 3300,
      IItems.铅锭, 2800,
      IItems.单晶硅, 2350,
      IItems.铱板, 1500,
      IItems.导能回路, 1900,
      IItems.陶钢, 900,
      IItems.肃正协议, 1
    )

    setAmmo()
  }

  fun setAmmo() {
    val laserColor1 = Color.valueOf("6569C9")
    val laserColor2 = Color.valueOf("8CA9E8")
    val laserColor3 = Color.valueOf("D1EFFF")

    newAmmo(object :LaserBulletType(3565f) {
      init {
        length = 800f
        width = 40f
        ammoMultiplier = 1f

        status = IStatus.湍能
        statusDuration = 120f
        buildingDamageMultiplier = 0.05f

        sideAngle = 22.5f

        colors = arrayOf(laserColor1, laserColor2, laserColor3)

        shootEffect = ParticleEffect().apply {
          line = true
          particles = 12
          lifetime = 20f
          length = 45f
          cone = 30f
          lenFrom = 6f
          lenTo = 6f
          strokeFrom = 3f
          strokeTo = 0f
          lightColor = laserColor2
          colorFrom = laserColor3
          colorTo = laserColor2
        }

        hitEffect = ParticleEffect().apply {
          line = true
          particles = 12
          lifetime = 20f
          length = 75f
          cone = 360f
          lenFrom = 6f
          lenTo = 6f
          strokeFrom = 3f
          strokeTo = 0f
          lightColor = laserColor2
          colorFrom = laserColor3
          colorTo = laserColor2
        }
      }
    }).apply {
      consume?.apply {
        time(1200f)
        power(165f)
        liquid(ILiquids.急冻液, 1.5f)
      }
    }
  }

}
